pkg/compact/compact.go

```diff
@@ -470,6 +481,102 @@ func (cg *Group) Resolution() int64 {
 	return cg.resolution
 }
 
+// metrics related to planning and compaction progress
+type ProgressMetrics struct {
+	NumberOfIterations    *prometheus.GaugeVec
+	NumberOfBlocksToMerge *prometheus.GaugeVec
+}
+
+// should return the results/metrics of the planning simulation
+type PlanSim interface {
+	ProgressCalculate(ctx context.Context, groups []*Group) error
+}
+
+type DefaultPlanSim struct {
+	planner Planner
+	*ProgressMetrics
+}
+
+func NewDefaultPlanSim(reg prometheus.Registerer, planner Planner) *DefaultPlanSim {
+	return &DefaultPlanSim{
+		planner: planner,
+		ProgressMetrics: &ProgressMetrics{
+			NumberOfIterations: promauto.With(reg).NewGaugeVec(prometheus.GaugeOpts{
+				Name: "thanos_number_of_iterations",
+				Help: "number of iterations to be done",
+			}, []string{"group"}),
+			NumberOfBlocksToMerge: promauto.With(reg).NewGaugeVec(prometheus.GaugeOpts{
+				Name: "thanos_number_of_blocks_planned",
+				Help: "number of blocks planned to be merged",
+			}, []string{"group"}),
+		},
+	}
+}
+
+func (ps *DefaultPlanSim) ProgressCalculate(ctx context.Context, groups []*Group) error {
+	iterations := 0
+	blocksToMerge := 0
+	groupCompactions := make(map[string]int, len(groups))
+	groupBlocks := make(map[string]int, len(groups))
+
+	hasPlan := true
+	for hasPlan {
+		for _, g := range groups {
+			if len(g.IDs()) == 1 {
+				continue
+			}
+
+			// parameter should be of type tsdb.BlockMeta.meta
+			plan, err := ps.planner.Plan(ctx, g.metasByMinTime)
+			if err != nil {
+				return errors.Wrapf(err, "could not plan")
+			}
+			if len(plan) == 0 {
+				continue
+			}
+			iterations++
+			groupCompactions[g.key] = iterations
+
+			// value type in map is struct{} - enpty struct consumes 0 bytes so prefered over bool
+			toRemove := make(map[ulid.ULID]struct{}, len(plan))
+			metas := make([]*tsdb.BlockMeta, 0, len(plan))
+			for _, p := range plan {
+				metas = append(metas, &p.BlockMeta)
+				toRemove[p.BlockMeta.ULID] = struct{}{}
+			}
+			g.deleteFromGroup(toRemove)
+
+			blocksToMerge += len(plan)
+			groupBlocks[g.key] = blocksToMerge
+
+			// remove 'plan' blocks from 'original metadata' - so that the remaining blocks can now be planned ?
+			// not required to modify originalMeta now
+
+			if len(g.metasByMinTime) == 0 {
+				hasPlan = false
```

Comment:
Why? If this group doesn't contain any block, then other groups might still have available plans. We cannot exit set it to false here.
Only if all groups don't have available plan can we exit the for loop